// devframe/dist/rpc-BlVCyk9W.mjs
import { t as hash } from "./hash-CIBte1yS.mjs";
import { i as logger } from "./serialization-CwO31axa.mjs";
import { DEVTOOLS_RPC_DUMP_DIRNAME } from "./constants.mjs";
//#region src/rpc/cache.ts
/**
* @experimental API is expected to change.
*/
var RpcCacheManager = class {
	cacheMap = /* @__PURE__ */ new Map();
	options;
	keySerializer;
	constructor(options) {
		this.options = options;
		this.keySerializer = options.keySerializer || ((args) => hash(args));
	}
	updateOptions(options) {
		this.options = {
			...this.options,
			...options
		};
	}
	cached(m, a) {
		const methodCache = this.cacheMap.get(m);
		if (methodCache) return methodCache.get(this.keySerializer(a));
	}
	apply(req, res) {
		const methodCache = this.cacheMap.get(req.m) || /* @__PURE__ */ new Map();
		methodCache.set(this.keySerializer(req.a), res);
		this.cacheMap.set(req.m, methodCache);
	}
	validate(m) {
		return this.options.functions.includes(m);
	}
	clear(fn) {
		if (fn) this.cacheMap.delete(fn);
		else this.cacheMap.clear();
	}
};
//#endregion
//#region src/rpc/handler.ts
async function getRpcResolvedSetupResult(definition, context) {
	if (definition.__resolved) return definition.__resolved;
	if (!definition.setup) return {};
	definition.__promise ??= Promise.resolve(definition.setup(context)).then((r) => {
		definition.__resolved = r;
		definition.__promise = void 0;
		return r;
	});
	return definition.__resolved ??= await definition.__promise;
}
async function getRpcHandler(definition, context) {
	if (definition.handler) return definition.handler;
	const result = await getRpcResolvedSetupResult(definition, context);
	if (!result.handler) throw logger.DF0024({ name: definition.name }).throw();
	return result.handler;
}
//#endregion
//#region src/rpc/collector.ts
var RpcFunctionsCollectorBase = class {
	context;
	definitions = /* @__PURE__ */ new Map();
	functions;
	_onChanged = [];
	constructor(context) {
		this.context = context;
		const definitions = this.definitions;
		const self = this;
		this.functions = new Proxy({}, {
			get(_, prop) {
				const definition = definitions.get(prop);
				if (!definition) return void 0;
				return getRpcHandler(definition, self.context);
			},
			has(_, prop) {
				return definitions.has(prop);
			},
			getOwnPropertyDescriptor(_, prop) {
				return {
					value: definitions.get(prop)?.handler,
					configurable: true,
					enumerable: true
				};
			},
			ownKeys() {
				return Array.from(definitions.keys());
			}
		});
	}
	register(fn, force = false) {
		if (this.definitions.has(fn.name) && !force) throw logger.DF0021({ name: fn.name }).throw();
		assertAgentJsonSerializable(fn);
		this.definitions.set(fn.name, fn);
		this._onChanged.forEach((cb) => cb(fn.name));
	}
	update(fn, force = false) {
		if (!this.definitions.has(fn.name) && !force) throw logger.DF0022({ name: fn.name }).throw();
		assertAgentJsonSerializable(fn);
		this.definitions.set(fn.name, fn);
		this._onChanged.forEach((cb) => cb(fn.name));
	}
	onChanged(fn) {
		this._onChanged.push(fn);
		return () => {
			const index = this._onChanged.indexOf(fn);
			if (index !== -1) this._onChanged.splice(index, 1);
		};
	}
	async getHandler(name) {
		return await getRpcHandler(this.definitions.get(name), this.context);
	}
	getSchema(name) {
		const definition = this.definitions.get(name);
		if (!definition) throw logger.DF0023({ name: String(name) }).throw();
		return {
			args: definition.args,
			returns: definition.returns
		};
	}
	has(name) {
		return this.definitions.has(name);
	}
	get(name) {
		return this.definitions.get(name);
	}
	list() {
		return Array.from(this.definitions.keys());
	}
};
function assertAgentJsonSerializable(fn) {
	if (fn.agent && fn.jsonSerializable !== true) throw logger.DF0019({ name: fn.name }).throw();
}
//#endregion
//#region ../../node_modules/.pnpm/yocto-queue@1.2.2/node_modules/yocto-queue/index.js
var Node = class {
	value;
	next;
	constructor(value) {
		this.value = value;
	}
};
var Queue = class {
	#head;
	#tail;
	#size;
	constructor() {
		this.clear();
	}
	enqueue(value) {
		const node = new Node(value);
		if (this.#head) {
			this.#tail.next = node;
			this.#tail = node;
		} else {
			this.#head = node;
			this.#tail = node;
		}
		this.#size++;
	}
	dequeue() {
		const current = this.#head;
		if (!current) return;
		this.#head = this.#head.next;
		this.#size--;
		if (!this.#head) this.#tail = void 0;
		return current.value;
	}
	peek() {
		if (!this.#head) return;
		return this.#head.value;
	}
	clear() {
		this.#head = void 0;
		this.#tail = void 0;
		this.#size = 0;
	}
	get size() {
		return this.#size;
	}
	*[Symbol.iterator]() {
		let current = this.#head;
		while (current) {
			yield current.value;
			current = current.next;
		}
	}
	*drain() {
		while (this.#head) yield this.dequeue();
	}
};
//#endregion
//#region ../../node_modules/.pnpm/p-limit@7.3.0/node_modules/p-limit/index.js
function pLimit(concurrency) {
	let rejectOnClear = false;
	if (typeof concurrency === "object") ({concurrency, rejectOnClear = false} = concurrency);
	validateConcurrency(concurrency);
	if (typeof rejectOnClear !== "boolean") throw new TypeError("Expected `rejectOnClear` to be a boolean");
	const queue = new Queue();
	let activeCount = 0;
	const resumeNext = () => {
		if (activeCount < concurrency && queue.size > 0) {
			activeCount++;
			queue.dequeue().run();
		}
	};
	const next = () => {
		activeCount--;
		resumeNext();
	};
	const run = async (function_, resolve, arguments_) => {
		const result = (async () => function_(...arguments_))();
		resolve(result);
		try {
			await result;
		} catch {}
		next();
	};
	const enqueue = (function_, resolve, reject, arguments_) => {
		const queueItem = { reject };
		new Promise((internalResolve) => {
			queueItem.run = internalResolve;
			queue.enqueue(queueItem);
		}).then(run.bind(void 0, function_, resolve, arguments_));
		if (activeCount < concurrency) resumeNext();
	};
	const generator = (function_, ...arguments_) => new Promise((resolve, reject) => {
		enqueue(function_, resolve, reject, arguments_);
	});
	Object.defineProperties(generator, {
		activeCount: { get: () => activeCount },
		pendingCount: { get: () => queue.size },
		clearQueue: { value() {
			if (!rejectOnClear) {
				queue.clear();
				return;
			}
			const abortError = AbortSignal.abort().reason;
			while (queue.size > 0) queue.dequeue().reject(abortError);
		} },
		concurrency: {
			get: () => concurrency,
			set(newConcurrency) {
				validateConcurrency(newConcurrency);
				concurrency = newConcurrency;
				queueMicrotask(() => {
					while (activeCount < concurrency && queue.size > 0) resumeNext();
				});
			}
		},
		map: { async value(iterable, function_) {
			const promises = Array.from(iterable, (value, index) => this(function_, value, index));
			return Promise.all(promises);
		} }
	});
	return generator;
}
function validateConcurrency(concurrency) {
	if (!((Number.isInteger(concurrency) || concurrency === Number.POSITIVE_INFINITY) && concurrency > 0)) throw new TypeError("Expected `concurrency` to be a number from 1 and up");
}
//#endregion
//#region src/rpc/validation.ts
/**
* Validates RPC function definitions.
* Action and event functions cannot have dumps (side effects should not be cached).
*
* @throws {Error} If an action or event function has a dump configuration
*/
function validateDefinitions(definitions) {
	for (const definition of definitions) {
		const type = definition.type || "query";
		if ((type === "action" || type === "event") && definition.dump) throw logger.DF0027({
			name: definition.name,
			type
		}).throw();
		if (definition.snapshot && type !== "query") throw logger.DF0028({
			name: definition.name,
			type
		}).throw();
	}
}
/**
* Validates a single RPC function definition.
*
* @throws {Error} If an action or event function has a dump configuration
*/
function validateDefinition(definition) {
	validateDefinitions([definition]);
}
//#endregion
//#region src/rpc/dump/error.ts
/**
* Normalize a thrown value into a plain object suitable for storage in
* a dump record. Preserves `message`, `name`, `cause`, and any own
* enumerable properties of an `Error` so consumers reading the dump can
* reconstruct a richer Error than just `{ message, name }`.
*
* Non-`Error` throws are wrapped as `{ name: 'Error', message: String(thrown) }`.
*/
function serializeDumpError(error) {
	return serializeWithSeen(error, /* @__PURE__ */ new WeakSet());
}
function serializeWithSeen(error, seen) {
	if (!(error instanceof Error)) return {
		name: "Error",
		message: String(error)
	};
	if (seen.has(error)) return {
		name: error.name,
		message: error.message
	};
	seen.add(error);
	const out = {
		name: error.name,
		message: error.message
	};
	const cause = error.cause;
	if (cause !== void 0) out.cause = cause instanceof Error ? serializeWithSeen(cause, seen) : cause;
	for (const key of Object.keys(error)) {
		if (key === "name" || key === "message" || key === "cause") continue;
		out[key] = error[key];
	}
	return out;
}
/**
* Inverse of {@link serializeDumpError}: rebuild a thrown `Error` from
* the plain object stored in a dump record. Preserves `cause`, restores
* the original `name`, and re-attaches any custom own properties.
*/
function reviveDumpError(stored) {
	const cause = stored.cause instanceof Error ? stored.cause : isPlainErrorShape(stored.cause) ? reviveDumpError(stored.cause) : stored.cause;
	const error = cause !== void 0 ? new Error(stored.message, { cause }) : new Error(stored.message);
	error.name = stored.name;
	for (const key of Object.keys(stored)) {
		if (key === "name" || key === "message" || key === "cause") continue;
		error[key] = stored[key];
	}
	return error;
}
function isPlainErrorShape(value) {
	return typeof value === "object" && value !== null && typeof value.message === "string" && typeof value.name === "string";
}
//#endregion
//#region src/rpc/dump/index.ts
function getDumpRecordKey(functionName, args) {
	return `${functionName}---${hash(args)}`;
}
function getDumpFallbackKey(functionName) {
	return `${functionName}---fallback`;
}
async function resolveGetter(valueOrGetter) {
	return typeof valueOrGetter === "function" ? await valueOrGetter() : valueOrGetter;
}
/**
* Collects pre-computed dumps by executing functions with their defined input combinations.
* Static functions without dump config automatically get `{ inputs: [[]] }`.
*
* @example
* ```ts
* const store = await dumpFunctions([greet], context, { concurrency: 10 })
* ```
*/
async function dumpFunctions(definitions, context, options = {}) {
	validateDefinitions(definitions);
	const concurrency = options.concurrency === true ? 5 : options.concurrency === false || options.concurrency == null ? 1 : options.concurrency;
	const store = {
		definitions: {},
		records: {}
	};
	const tasksResolutions = definitions.map((definition) => async () => {
		if (definition.type === "event" || definition.type === "action") return;
		const setupResult = definition.setup ? await Promise.resolve(definition.setup(context)) : {};
		const handler = setupResult.handler || definition.handler;
		if (!handler) throw logger.DF0024({ name: definition.name }).throw();
		let dump = setupResult.dump ?? definition.dump;
		if (!dump && definition.type === "static") dump = { inputs: [[]] };
		if (!dump && definition.snapshot) dump = async (_ctx, h) => {
			const output = await Promise.resolve(h(...[]));
			return {
				records: [{
					inputs: [],
					output
				}],
				fallback: output
			};
		};
		if (!dump) return;
		if (typeof dump === "function") dump = await Promise.resolve(dump(context, handler));
		store.definitions[definition.name] = {
			name: definition.name,
			type: definition.type
		};
		return {
			handler,
			dump,
			definition
		};
	});
	let functionsToDump = [];
	if (concurrency <= 1) for (const task of tasksResolutions) {
		const resolution = await task();
		if (resolution) functionsToDump.push(resolution);
	}
	else {
		const limit = pLimit(concurrency);
		functionsToDump = (await Promise.all(tasksResolutions.map((task) => limit(task)))).filter((x) => !!x);
	}
	const dumpTasks = [];
	for (const { definition, handler, dump } of functionsToDump) {
		const { inputs, records, fallback } = dump;
		if (records) for (const record of records) {
			const recordKey = getDumpRecordKey(definition.name, record.inputs);
			store.records[recordKey] = record;
		}
		if ("fallback" in dump) {
			const fallbackKey = getDumpFallbackKey(definition.name);
			store.records[fallbackKey] = {
				inputs: [],
				output: fallback
			};
		}
		if (inputs) for (const input of inputs) dumpTasks.push(async () => {
			const recordKey = getDumpRecordKey(definition.name, input);
			try {
				const output = await Promise.resolve(handler(...input));
				store.records[recordKey] = {
					inputs: input,
					output
				};
			} catch (error) {
				store.records[recordKey] = {
					inputs: input,
					error: serializeDumpError(error)
				};
			}
		});
	}
	if (concurrency <= 1) for (const task of dumpTasks) await task();
	else {
		const limit = pLimit(concurrency);
		await Promise.all(dumpTasks.map((task) => limit(task)));
	}
	return store;
}
/**
* Creates a client that serves pre-computed results from a dump store.
* Uses argument hashing to match calls to stored records.
*
* @example
* ```ts
* const client = createClientFromDump(store)
* await client.greet('Alice')
* ```
*/
function createClientFromDump(store, options = {}) {
	const { onMiss } = options;
	return new Proxy({}, {
		get(_, functionName) {
			if (!(functionName in store.definitions)) throw logger.DF0025({ name: functionName }).throw();
			return async (...args) => {
				const recordKey = getDumpRecordKey(functionName, args);
				const recordOrGetter = store.records[recordKey];
				if (recordOrGetter) {
					const record = await resolveGetter(recordOrGetter);
					if (record.error) throw reviveDumpError(record.error);
					if (typeof record.output === "function") return await record.output();
					return record.output;
				}
				onMiss?.(functionName, args);
				const fallbackKey = getDumpFallbackKey(functionName);
				if (fallbackKey in store.records) {
					const fallbackOrGetter = store.records[fallbackKey];
					const fallbackRecord = await resolveGetter(fallbackOrGetter);
					if (fallbackRecord && typeof fallbackRecord.output === "function") return await fallbackRecord.output();
					if (fallbackRecord) return fallbackRecord.output;
				}
				throw logger.DF0026({
					name: functionName,
					args: JSON.stringify(args)
				}).throw();
			};
		},
		has(_, functionName) {
			return functionName in store.definitions;
		},
		ownKeys() {
			return Object.keys(store.definitions);
		},
		getOwnPropertyDescriptor(_, functionName) {
			return functionName in store.definitions ? {
				configurable: true,
				enumerable: true,
				value: void 0
			} : void 0;
		}
	});
}
/**
* Filters function definitions to only those with dump definitions.
* Note: Only checks the definition itself, not setup results.
*/
function getDefinitionsWithDumps(definitions) {
	return definitions.filter((def) => def.dump !== void 0);
}
//#endregion
//#region src/rpc/dump/static.ts
function makeDumpKey(name) {
	return encodeURIComponent(name.replaceAll(":", "~"));
}
function makeStaticPath(name) {
	return `${DEVTOOLS_RPC_DUMP_DIRNAME}/${makeDumpKey(name)}.static.json`;
}
function makeQueryRecordPath(name, hash) {
	return `${DEVTOOLS_RPC_DUMP_DIRNAME}/${makeDumpKey(name)}.record.${hash}.json`;
}
function makeQueryFallbackPath(name) {
	return `${DEVTOOLS_RPC_DUMP_DIRNAME}/${makeDumpKey(name)}.fallback.json`;
}
async function resolveRecord(record) {
	return typeof record === "function" ? await record() : record;
}
async function collectStaticRpcDump(definitions, context) {
	const manifest = {};
	const files = {};
	for (const definition of definitions) {
		const type = definition.type ?? "query";
		const serialization = definition.jsonSerializable === true ? "json" : "structured-clone";
		if (type === "static") {
			const handler = await getRpcHandler(definition, context);
			const path = makeStaticPath(definition.name);
			files[path] = {
				serialization,
				fnName: definition.name,
				data: { output: await Promise.resolve(handler()) }
			};
			manifest[definition.name] = {
				type: "static",
				path,
				serialization
			};
			continue;
		}
		if (type !== "query") continue;
		const store = await dumpFunctions([definition], context);
		if (!(definition.name in store.definitions)) continue;
		const queryEntry = {
			type: "query",
			records: {},
			serialization
		};
		const prefix = `${definition.name}---`;
		for (const [recordKey, recordOrGetter] of Object.entries(store.records)) {
			if (!recordKey.startsWith(prefix)) continue;
			const key = recordKey.slice(prefix.length);
			const record = await resolveRecord(recordOrGetter);
			if (key === "fallback") {
				const path = makeQueryFallbackPath(definition.name);
				files[path] = {
					serialization,
					fnName: definition.name,
					data: record
				};
				queryEntry.fallback = path;
			} else {
				const path = makeQueryRecordPath(definition.name, key);
				files[path] = {
					serialization,
					fnName: definition.name,
					data: record
				};
				queryEntry.records[key] = path;
			}
		}
		if (!Object.keys(queryEntry.records).length && !queryEntry.fallback) continue;
		manifest[definition.name] = queryEntry;
	}
	return {
		manifest,
		files
	};
}
//#endregion
export { reviveDumpError as a, validateDefinitions as c, getRpcResolvedSetupResult as d, RpcCacheManager as f, getDefinitionsWithDumps as i, RpcFunctionsCollectorBase as l, createClientFromDump as n, serializeDumpError as o, dumpFunctions as r, validateDefinition as s, collectStaticRpcDump as t, getRpcHandler as u };
