// @vitejs/devtools/dist/cli-commands.js
import { r as logger } from "./constants-CwdrFfgU.js";
import { normalizeHttpServerUrl } from "devframe/node";
import { colors } from "devframe/utils/colors";
import { open } from "devframe/utils/open";
import { resolve } from "pathe";
import { DEVTOOLS_MOUNT_PATH } from "@vitejs/devtools-kit/constants";
//#region src/node/cli-commands.ts
async function start(options) {
	const { host } = options;
	const { getPort } = await import("./dist-CAx5WPY9.js");
	const port = await getPort({
		host,
		port: options.port == null ? void 0 : +options.port,
		portRange: [9999, 15e3]
	});
	const { startStandaloneDevTools } = await import("./standalone-TG1krjts.js");
	const { createDevToolsMiddleware } = await import("./server-DRN3It5u.js");
	const devtools = await startStandaloneDevTools({ cwd: options.root });
	const { h3 } = await createDevToolsMiddleware({
		cwd: devtools.config.root,
		websocket: {
			host,
			https: false
		},
		context: devtools.context
	});
	const { createServer } = await import("node:http");
	const { defineHandler, H3, sendRedirect, toNodeHandler } = await import("h3");
	const { mountStaticHandler } = await import("devframe/utils/serve-static");
	const app = new H3();
	for (const { baseUrl, distDir } of devtools.context.views.buildStaticDirs) mountStaticHandler(app, baseUrl, distDir);
	app.use(DEVTOOLS_MOUNT_PATH, h3);
	app.use("/", defineHandler((event) => sendRedirect(event, DEVTOOLS_MOUNT_PATH, 302)));
	createServer(toNodeHandler(app)).listen(port, host, async () => {
		const url = normalizeHttpServerUrl(host, port);
		console.log(colors.green`${"⬢"} Vite DevTools started at`, colors.green(url), "\n");
		if (options.open) await open(url);
	});
}
async function build(options) {
	console.log(colors.cyan`${"⬢"} Building static Vite DevTools...`);
	const { startStandaloneDevTools } = await import("./standalone-TG1krjts.js");
	const devtools = await startStandaloneDevTools({
		cwd: options.root,
		config: options.config
	});
	const outDir = resolve(devtools.config.root, options.outDir);
	const { buildStaticDevTools } = await import("./build-static-Bt-O93wY.js");
	await buildStaticDevTools({
		context: devtools.context,
		outDir
	});
	logger.DTK0010().log();
}
//#endregion
export { build, start };
